File: Array&String/sort_array_by_parity2.py
class Solution:
    def sortArrayByParityII(self, nums: list[int]) -> list[int]:
        if len(nums) == 0:
            return nums
        writeIdx = 0
        for readIdx in range(len(nums)):
            if nums[readIdx] % 2 == 0:
                nums[readIdx], nums[writeIdx] = nums[writeIdx], nums[readIdx]
                writeIdx += 1
        j = 1
        for i in range(len(nums)):
            if nums[i] & 1:
                nums[i], nums[j] = nums[j], nums[i]
                j += 2
        return nums
    
    # Method 2
    def sortArrayByParityII(self, a):
        i = 0 # pointer for even misplaced
        j = 1 # pointer for odd misplaced
        sz = len(a)
        
        # invariant: for every misplaced odd there is misplaced even
        # since there is just enough space for odds and evens

        while i < sz and j < sz:
            if a[i] % 2 == 0:
                i += 2
            elif a[j] % 2 == 1:
                j += 2
            else:
                # a[i] % 2 == 1 AND a[j] % 2 == 0
                a[i],a[j] = a[j],a[i]
                i += 2
                j += 2

        return a

File: Array&String/CPP/sort_array_by_parity2.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> sortArrayByParityII(vector<int>& nums) {
        if (nums.size() <= 1) return nums;
        int i = 0;
        for (int j = 0; j < nums.size(); j++) {
            if (nums[j] % 2 == 0) {
                int temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
                i++;
            }
        }

        int j = 1;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] % 2 == 1) {
                int temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
                j += 2;
            }
        }
    return nums;
    }

    // method 2
    vector<int> sortArrayByParityII(vector<int>& nums) {
        int i = 0, j = 1, n = nums.size();
        while (i < n && j < n) {
            if (nums[i] % 2 == 0) i += 2;
            else if (nums[j] % 2 == 1) j += 2;
            else swap(nums[i], nums[j]);
        }
        return nums;
    }
};